pkg/csplugin/broker.go

```diff
@@ -0,0 +1,419 @@
+	}
+	pb.watcher = PluginWatcher{}
+	pb.watcher.Init(pb.pluginConfigByName, pb.alertsByPluginName)
+	return nil
+
+}
+
+func (pb *PluginBroker) Kill() {
+	for _, kill := range pb.pluginKillMethods {
+		kill()
+	}
+}
+
+func (pb *PluginBroker) Run(tomb *tomb.Tomb) {
+	pb.watcher.Start(tomb)
+	for {
+		select {
+		case profileAlert := <-pb.PluginChannel:
+			pb.addProfileAlert(profileAlert)
+
+		case pluginName := <-pb.watcher.PluginEvents:
+			// this can be ran in goroutine, but then locks will be needed
+			pluginMutex.Lock()
+			tmpAlerts := pb.alertsByPluginName[pluginName]
+			pb.alertsByPluginName[pluginName] = make([]*models.Alert, 0)
+			pluginMutex.Unlock()
+			go func() {
+				if err := pb.pushNotificationsToPlugin(pluginName, tmpAlerts); err != nil {
+					log.WithField("plugin:", pluginName).Error(err)
+				}
+			}()
+
+		case <-tomb.Dying():
+			log.Info("killing all plugins")
+			pb.Kill()
+			return
+		}
+	}
+}
+func (pb *PluginBroker) addProfileAlert(profileAlert ProfileAlert) {
+	for _, pluginName := range pb.profileConfigs[profileAlert.ProfileID].Notifications {
+		if _, ok := pb.pluginConfigByName[pluginName]; !ok {
+			log.Errorf("plugin %s is not configured properly.", pluginName)
+			continue
+		}
+		pluginMutex.Lock()
+		pb.alertsByPluginName[pluginName] = append(pb.alertsByPluginName[pluginName], profileAlert.Alert)
+		pluginMutex.Unlock()
+		pb.watcher.Inserts <- pluginName
+	}
+}
+func (pb *PluginBroker) profilesContainPlugin(pluginName string) bool {
+	for _, profileCfg := range pb.profileConfigs {
+		for _, name := range profileCfg.Notifications {
+			if pluginName == name {
+				return true
+			}
+		}
+	}
+	return false
+}
+func (pb *PluginBroker) loadConfig(path string) error {
+	files, err := listFilesAtPath(path)
+	if err != nil {
+		return err
+	}
+	for _, configFilePath := range files {
+		if !strings.HasSuffix(configFilePath, ".yaml") && !strings.HasSuffix(configFilePath, ".yml") {
+			continue
+		}
+
+		pluginConfigs, err := parsePluginConfigFile(configFilePath)
+		if err != nil {
+			return errors.Wrapf(err, "got error while parsing %s", configFilePath)
+		}
+		for _, pluginConfig := range pluginConfigs {
+			if !pb.profilesContainPlugin(pluginConfig.Name) {
+				continue
+			}
+			setRequiredFields(&pluginConfig)
+			if _, ok := pb.pluginConfigByName[pluginConfig.Name]; ok {
+				log.Warnf("several configs for notification %s found  ", pluginConfig.Name)
+			}
+			pb.pluginConfigByName[pluginConfig.Name] = pluginConfig
+		}
+	}
+	err = pb.verifyPluginConfigsWithProfile()
+	return err
+}
+
+func (pb *PluginBroker) verifyPluginConfigsWithProfile() error {
+	for _, profileCfg := range pb.profileConfigs {
+		for _, pluginName := range profileCfg.Notifications {
+			if _, ok := pb.pluginConfigByName[pluginName]; !ok {
+				return fmt.Errorf("config file for plugin %s not found", pluginName)
+			}
+		}
+	}
+	return nil
+}
+
+func (pb *PluginBroker) loadPlugins(path string) error {
```

Comment:
maybe we want to error.Wrap returned errors here to provide more context on the error ?

Reply:
no need because we are wrapping it where it is catched, I think wrapping again here won't do much good